rs/utils/src/l2.rs

```diff
@@ -1,13 +1,100 @@
+use std::{
+    ops::Mul,
+    simd::{f32x4, f32x8, num::SimdFloat},
+};
+
 use crate::DistanceCalculator;
 
 pub struct L2DistanceCalculator {}
 
-impl DistanceCalculator for L2DistanceCalculator {
-    fn calculate(&self, a: &[f32], b: &[f32]) -> f32 {
+impl L2DistanceCalculator {
+    pub fn new() -> Self {
+        Self {}
+    }
+
+    #[allow(dead_code)]
+    fn calculate_non_simd(&self, a: &[f32], b: &[f32]) -> f32 {
         let mut dist = 0.0;
         for i in 0..a.len() {
             dist += (a[i] - b[i]).powi(2);
         }
         dist.sqrt()
     }
 }
+
+impl DistanceCalculator for L2DistanceCalculator {
+    fn calculate(&self, a: &[f32], b: &[f32]) -> f32 {
+        let mut i = 0;
+        let mut dist_8 = f32x8::splat(0.0);
+        let mut dist_4 = f32x4::splat(0.0);
+        let mut dist = 0.0;
+        while i + 8 <= a.len() {
+            let a = f32x8::from_array(a[i..i + 8].try_into().unwrap());
```

Comment:
I believe we can replace this with `let a = f32x8::from_slice(&a[i..]);`, right?

Reply:
you need a[i..i+8` actually. this function expects a fixed sized slice

Reply:
from https://doc.rust-lang.org/std/simd/prelude/struct.Simd.html#method.from_slice, it seems like `from_slice` will always take 8 elements for f32x8-typed vector, right? And the `while` loop condition ensures that we have at least 8 elements in the slice at this point. Am I missing something?